Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1869,47 +1937,55 @@ public partial class ModuleResolver {
-            if (arg is LiteralExpr lit && lit.Value is string s) {
-              // Get the NativeType for "s"
-              foreach (var nativeT in NativeTypes) {
-                if (nativeT.Name == s) {
-                  nativeTypeChoices.Add(nativeT);
-                  goto FoundNativeType;
-                }
-              }
+      // Look at the :nativeType attribute, if any, to determine preferences.
+      // A return of false means: don't use native type. (nativeTypes returns as null)
+      // A return of null means: no preference. (nativeTypes returns as null)
+      // A return of true means: make sure a native type is used. Furthermore,
+      //   * if nativeTypes is null, then no particular preference about which native type is picked
+      //   * if nativeTypes is non-null, then the choices of native types are restricted to those.
+      bool? ReadAttributesForNativePreferences(Attributes attributes, [CanBeNull] out List<NativeType> nativeTypes) {
+        nativeTypes = null;
+        var args = Attributes.FindExpressions(attributes, "nativeType");
+        if (args == null) {
+          // There was no :nativeType attribute
+          return null;
+        }
+        if (args.Count == 0) {
+          // {:nativeType}
+          return true;
+        }
+        if (args[0] is LiteralExpr { Value: bool and var boolValue }) {
+          return boolValue;
+        }
+
+        var choices = new List<NativeType>();
+        foreach (var arg in args) {
+          if (arg is LiteralExpr { Value: string s }) {
+            // Get the NativeType for "s"
+            var nativeType = NativeTypes.Find(nativeType => nativeType.Name == s);
+            if (nativeType == null) {
               reporter.Error(MessageSource.Resolver, dd, ":nativeType '{0}' not known", s);
-              return;
-            FoundNativeType:;
-            } else {
-              reporter.Error(MessageSource.Resolver, arg, "unexpected :nativeType argument");
-              return;
+              return false;
             }
+            choices.Add(nativeType);
+          } else {
+            reporter.Error(MessageSource.Resolver, arg, "unexpected :nativeType argument");
+            return false;
           }
         }
+        nativeTypes = choices;
+        return true;
+      }
+
+      // Look at the :nativeType attribute, if any
+      var hasNativeTypePreference = ReadAttributesForNativePreferences(dd.Attributes, out var nativeTypeChoices);
+      if (hasNativeTypePreference == false) {
+        // don't use native type
+        return;
+      }
+      if (hasNativeTypePreference == true && dd.BaseType.NormalizeToAncestorType() is not (IntType or BitvectorType)) {
+        reporter.Error(MessageSource.Resolver, dd, ":nativeType can only be used on a newtype based on integers or bitvectors");
+        return;
       }
 
       // Figure out the variable and constraint.  Usually, these would be just .Var and .Constraint, but
```

Comment:
I don't get this. When `dd.Constraint` is set, does it include the transitive constraints from base types? If so, why is `dd.Constraint` not also set even when `ddVar == null`, which would make this loop obsolete?

Reply:
No, `dd.Var` and `dd.Constraint` are the syntactic components from the program text. So, for

``` dafny
newtype MyConstrainedInt = x: int | 0 <= x < 100
```

`dd.Var` is `x` and `dd.Constraint` is `0 <= x < 100`, and for

``` dafny
newtype MyInt = int
```

both `dd.Var` and `dd.Constraint` are `null`.

Reply:
I don't get this loop. `FigureOutNativeTypeForIntegerNewtype` calls `DiscoverAllBounds_SingleVar` which also traverses the chain to collect constraints. Why do we also need to traverse here? I suggest removing this loop and only passing `NewType dd` to `FigureOutNativeTypeForIntegerNewtype` and `FigureOutNativeTypeForBitvectorNewtype`